Proxy dap requests to upstream clients

Anthony-Eid opened the pull request:
### Release Notes
This PR sets up the foundation to enable remote clients to send dap requests to any debugging sessions they're apart of. I also enabled remote clients to use all debug panel item buttons except for ignore breakpoints.

I have left out implementing variable/scope and console evaluation requests for this PR because I'll have to update the collab DB to add those requests and that's outside the scope of this PR.

### TODO
- [x] Step Over
- [x] Step In
- [x] Step Out
- [x] Step Back
- [x] Pause
- [x] Continue
- [x] Disconnect
- [x] Sync thread state on Continue/Pause button
- [x] Add integration test

### Preview

https://github.com/user-attachments/assets/36843885-4f5b-4899-8b00-c8d208ea1eda

Anthony-Eid committed:
WIP Start work to send all dap client requests with request_dap

Anthony-Eid committed:
Continue work on converting dap client requests

Anthony-Eid commented:
@RemcoSmitsDev This is still a WIP, but I would appreciate any feedback you have

RemcoSmitsDev commented on crates/project/src/dap_store.rs:
```diff
@@ -1164,38 +1156,42 @@ impl DapStore {
+
+            let task = self.request_dap::<SetVariable>(client_id, arguments, cx);
+
+            cx.background_executor().spawn(async move {
```
I think this wrapping it inside a background task is not needed, as you already do this inside the request_dap method.

Anthony-Eid replied:
I tried doing this

```rust
task.map(|res| res.map(|response| response.variables)) 
```

before, but the return type wasn't the same as the function. I do think you're right that we shouldn't do create two tasks, I'm going to change the function's return type instead. Thanks for the feedback!!

RemcoSmitsDev commented on crates/project/src/dap_store.rs:
```diff
@@ -967,14 +967,44 @@ impl DapStore {
+        cx: &mut ModelContext<Self>,
+    ) -> Task<Result<R::Response>>
+    where
+        <R as dap::requests::Request>::Response: 'static,
```
Do we have to specify the lifetime here? Not sure if this is really needed.

Anthony-Eid replied:
Without it the compiler returns an error when we try to build.

I think it's because we're moving that type into it's own thread and it needs to have ownership over it's fields to guarantee that they're still exist in the Task.

RemcoSmitsDev commented:
I think you are going in the right direction, one thing that came to my mind is that we now also have to sync the capabilities. Because we are using this to determine if we are sending a specific request or not, or use another request that is supported.

Anthony-Eid commented:
> I think you are going in the right direction, one thing that came to my mind is that we now also have to sync the capabilities. Because we are using this to determine if we are sending a specific request or not, or use another request that is supported.

Thank you for catching this!! I didn't even think about that

Anthony-Eid committed:
Merge branch 'debugger' into dap-proto-request

Anthony-Eid committed:
WIP setup dap command for proto dap requests

Anthony-Eid committed:
WIP dap command

Co-authored-by: Remco Smits <djsmits12@gmail.com>

Anthony-Eid committed:
revert "WIP dap command"

This reverts commit fd2a6832b667aa23caf588c3ab55243319bc1654.

Co-authored-by: Remco Smits <djsmits12@gmail.com>

Anthony-Eid committed:
More WIP with Dap Command trait
Co-authored-by: Remco Smits <djsmits12@gmail.com>

Anthony-Eid committed:
Get step over command to work with remote dap clients

Co-authored-by: Remco Smits <djsmits12@gmail.com>

Anthony-Eid committed:
Fix thread status not being set to stop on remote debug panel items

Anthony-Eid committed:
Create a inner wrapper type to use for dap remote step requests

Anthony-Eid committed:
Implement step in,back,out for remote debugger sessions

Anthony-Eid committed:
Add Continue Command

Anthony-Eid committed:
Merge branch 'debugger' into dap-proto-request

Anthony-Eid committed:
Add more dap command impls

TerminateThreads, Pause, and Disconnect. As well as a shutdown session request downstream clients can send to host

Anthony-Eid committed:
Add Disconnect & Terminate dap command impls

Anthony-Eid committed:
Merge branch 'debugger' into dap-proto-request

Anthony-Eid committed:
Add basic dap proxy test over collab

Anthony-Eid committed:
Fix clippy error

RemcoSmitsDev commented on crates/project/src/dap_store.rs:
```diff
@@ -1159,31 +1194,27 @@ impl DapStore {
-        };
-
         let capabilities = self.capabilities_by_id(client_id);
+        if capabilities.supports_step_back.unwrap_or_default() {
```
```suggestion
        if !capabilities.supports_step_back.unwrap_or_default() {
```

RemcoSmitsDev commented on crates/project/src/dap_store.rs:
```diff
@@ -1373,36 +1385,49 @@ impl DapStore {
+
+            let task = self.request_dap(client_id, command, cx);
+
+            cx.background_executor()
```
Spawning the background thread is not needed as the request_dap already does this for us

RemcoSmitsDev commented on crates/debugger_ui/src/debugger_panel_item.rs:
```diff
@@ -251,6 +251,7 @@ impl DebugPanelItem {
         });
 
         self.active_thread_item = ThreadItem::from_proto(state.active_thread_item());
+        self.update_thread_state_status(ThreadStatus::Stopped, cx);
```
Why do we have to do this? We already update the status few lines a both.

RemcoSmitsDev commented on crates/collab/src/tests/debug_panel_tests.rs:
```diff
@@ -299,3 +299,271 @@ async fn test_active_debug_panel_item_set_on_join_project(
+    cx_a.run_until_parked();
+    cx_b.run_until_parked();
+
+    // assert we don't have a debug panel item anymore because the client shutdown
```
Can we remove this from the test, because it's not what we want to test here. 
So its more clear if you see the test what it does.

RemcoSmitsDev reviewed (changes requested):
Great work!! I have a few small points that I would like to change.

Anthony-Eid committed:
Start work on syncing breakpoint thread status

Co-authored-by: Remco Smits <djsmits12@gmail.com>
Co-authored-by: Carter Canedy <cartercanedy42@gmail.com>

Anthony-Eid committed:
WIP Fix thread status not syncing

Anthony-Eid committed:
Add thread state model's to remote debug panels when setting panel items

Anthony-Eid committed:
Sync thread state on step out command

Anthony-Eid commented:
Note for anyone seeing this in the future, Remco and I talked about the review in person that's why non of his comments have been addressed
